### PR TITLE
attribute-ordering: Add attribute ordering check

### PR DESCRIPTION
**Pull request by jtojnar**
Based on https://discourse.nixos.org/t/document-attribute-ordering-in-package-expressions/4887.

cc @worldofpeace

### CODE REVIEW

File: overlays/attribute-ordering.nix [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+      lib.forEach sideBySide ({ fst, snd }:
+        let
+          fstInfo = preferredOrdering.${fst};
+          sndInfo = preferredOrdering.${snd};
+        in {
+          cond = fstInfo.order > sndInfo.order;
+          msg = ''
+            The ${lib.optionalString (fstInfo.group != null) "${fstInfo.group}, including the "}attribute “${fst}” should preferably come before ${lib.optionalString (sndInfo.group != null) "${sndInfo.group}’ "}“${snd}” attribute in the expression.
+
+            See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/attribute-ordering.md
```
[the remark this answers is not in the document]
Comment (worldofpeace) [reply overlays/attribute-ordering.nix:49]: Out of scope here, though.

**Comment by worldofpeace** [comment]
The ordering documented here is exactly as what seems to be, i guess logically, the preferred order.

File: overlays/attribute-ordering.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (jtojnar) [reply overlays/attribute-ordering.nix:49]: Opened https://github.com/jtojnar/nixpkgs-hammering/issues/7

File: explanations/attribute-ordering.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  };
+
+  meta = with stdenv.lib; {
+    description = "";
+    longDescription = ''
+    '';
+    homepage = "";
+    license = licenses.mit;
+    platforms = platforms.unix;
+    maintainers = [];
```
[the remark this answers is not in the document]
Comment (jtojnar) [reply explanations/attribute-ordering.md]: Oh, yeah. Long style everywhere that changes more than rarely. I probably would not bother with `outputs` unless there are more than five, or some of them need explanation comment.

**Comment by jtojnar** [comment]
I wanted to add `meta` orderings and `output` orderings but was not sure if it should all be in `attribute-ordering` linter. The former maybe, the latter probably not.

`*Inputs` will be hard, as the Nix code is not aware what are the value names, only the values. I guess we could approximate it by taking `drv.name` but it would be far from perfect. Ideally, we could also access Nix AST but I have not yet came up with a battle plan (will need to choose between rnix and hnix).

I am still not 100% on `mesonFlags` alphabetization, there are some benefits to having the same order as `meson_options.txt`, yet alphabetization makes sense too. I will think on it some more.

**Pull request merged**